pontoon/static/js/utils/data.js

```diff
@@ -0,0 +1,90 @@
+
+import React from 'react';
+
+import {ajax} from 'utils/ajax';
+
+
+export class DataManager {
+
+    constructor (component, data) {
+        this.component = component;
+        this.handleResponse = this.handleResponse.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+        this.refreshData = this.refreshData.bind(this);
+        this.initialData = data;
+    }
+
+    get data () {
+        const {data} = this.component.state;
+        return data;
+    }
+
+    get errors () {
+        const {errors} = this.component.state;
+        return errors;
+    }
+
+    get api () {
+        return this.component.props.api;
+    }
+
+    get requestMethod () {
+        return this.component.props.requestMethod || 'get';
+    }
+
+    get submitMethod () {
+        return this.component.props.submitMethod || this.requestMethod;
+    }
+
+    async refreshData (params) {
+        if (this.initialData) {
+            this.component.setState({data: this.initialData, errors: []});
+            this.initialData = undefined;
+            return
+        }
+        return this.handleResponse(
+            await ajax[this.requestMethod](this.api, params));
+    }
+
+    async handleResponse (response) {
+        const {status} = response;
+        const json = await response.json();
+        if (status === 200) {
+            const {data} = json;
+            this.component.setState({data, errors: {}});
+            return;
+        }
+        const {errors} = json;
+        this.component.setState({errors});
+    }
+
+    async handleSubmit (params) {
+        return this.handleResponse(
+            await ajax[this.submitMethod](this.api, params));
```

Comment:
Why not use `ajax.fetch` here? It has some built-in protections that might be convenient. It would also make it a bit less difficult to track down code paths while debugging.

Reply:
ill update